### PR TITLE
document store: add JSON type and codec.

### PR DESCRIPTION
**Pull request by hicqu**
The JSON binary representation is same with MySQL 5.7. we prefer
this not bson because 1) bson only supports JSON compound types
but not JSON primitive types, and 2) this representation is better
than bson on random access.

This PR now can support these two statements:
```
CREATE TABLE t (a json_field);
INSERT INTO t (a) values ('{"a": "b"}');
SELECT * FROM t;
```

### CODE REVIEW

File: server/util.go [hunk 1]
```diff
@@ -348,6 +348,9 @@ func dumpTextValue(mysqlType uint8, value types.Datum) ([]byte, error) {
 		return hack.Slice(value.GetMysqlEnum().String()), nil
 	case types.KindMysqlSet:
 		return hack.Slice(value.GetMysqlSet().String()), nil
+	case types.KindMysqlJSON:
+		j := value.GetMysqlJSON()
+		return hack.Slice(j.DumpToString()), nil
```
Comment (hanfei1991) [inline server/util.go]: use value.GetMysqlJSON().DumpToString() ?
btw, why not use String() but DumpToString()

File: util/types/json/serde.go [hunk 2]
```diff
@@ -0,0 +1,418 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (hanfei1991) [inline util/types/json/serde.go]: May be serdes.go is a better name ?
See https://en.wikipedia.org/wiki/SerDes

File: util/types/json/serde.go [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+	data-length ::= uint8*	// If the high bit of a byte is 1, the length
+                            // field is continued in the next byte,
+							// otherwise it is the last byte of the length
+							// field. So we need 1 byte to represent
+							// lengths up to 127, 2 bytes to represent
+							// lengths up to 16383, and so on...
+*/
+func serialize(in interface{}) (out []byte, err error) {
+	var typeCode byte
+	if typeCode, err = jsonTypeCode(in); err != nil {
```
Comment (hanfei1991) [inline util/types/json/serde.go]: typeCode, err := ...
if err != nil {...}

This is clearer

File: util/types/json/serde.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+		return
+	}
+	var bytesBuffer = bytes.NewBuffer(nil)
+	bytesBuffer.WriteByte(typeCode)
+	push(bytesBuffer, in)
```
Comment (hanfei1991) [inline util/types/json/serde.go]: It seems before everytime we push `in` , we need to write typecode. How about do it in `push` function to save a type check.
You can write a bench for `ser` and `des` to prove this.

File: util/types/json/impl.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		return ErrInvalidJSONText.GenByArgs("The document is empty")
+	}
+	err = bson.UnmarshalJSON([]byte(s), &j.json)
+	if err != nil {
+		return ErrInvalidJSONText.GenByArgs(err)
+	}
+	return err
+}
+
+func (j *jsonImpl) DumpToString() string {
```
Comment (hanfei1991) [inline util/types/json/impl.go]: What's the different between DumpToString and Serialize ? Why we return DumpToString to MySQL Client?

File: util/types/json/json.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	// ParseFromString parses a json from string.
+	ParseFromString(s string) error
+	// DumpToString dumps itself to string.
+	DumpToString() string
+	// Serialize means serialize itself into bytes.
+	Serialize() []byte
+	// Deserialize means deserialize a json from bytes.
+	Deserialize(bytes []byte)
+
+	// Extract is used for json_extract function.
```
Comment (hanfei1991) [inline util/types/json/json.go]: This PR is too long to review, maybe you can implement `Extract` and `Unquote` next time.

File: util/types/json/json.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"bytes"
+
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/terror"
+)
+
+// JSON is for MySQL JSON type.
+type JSON interface {
```
Comment (hanfei1991) [inline util/types/json/json.go]: Just make JSON a interface{} or map[string]interface{} ?
I am not very sure that if a number is still a complete JSON ?

File: util/types/json/impl.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/juju/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var _ JSON = &jsonImpl{}
+
+type jsonImpl struct {
+	json interface{}
+}
+
```
Comment (hanfei1991) [inline util/types/json/impl.go]: add comments for every exported function.

File: util/types/json/json.go (hunk 7 above)
Comment (hicqu) [inline util/types/json/json.go]: yes, all of number, string, nil, true/false are valid JSON primitive types. you can run this in MySQL:
```
insert into table t (json_field) values ('true');
```

File: server/util.go (hunk 1 above)
Comment (hicqu) [inline server/util.go]: DumpToString is a revert of ParseFromString. I think it's more clearly.

File: util/types/json/serde.go (hunk 4 above)
Comment (hicqu) [inline util/types/json/serde.go]: at line 196, we push(value) without prefix typecode. so we should seperate push prefix typecode and value.

File: util/types/json/json.go (hunk 7 above)
Comment (hanfei1991) [inline util/types/json/json.go]: Got it. Can you use a JSONObject interface, Let Bool, Int***, map[string]JSONObject, []JSONObject implement this interface.
This interface contains GetType(), GetBytes(), etc...

Comment (hanfei1991) [inline util/types/json/json.go]: type int16 Int16
func (t Int16) GetType() {return ...} Is better

File: util/types/json/impl.go (hunk 5 above)
Comment (hicqu) [inline util/types/json/impl.go]: Serialize serialize JSON to a binary format, it's good for store and mantipulate, but MySQL client can only retreive string.

File: util/types/json/json.go (hunk 7 above)
Comment (hicqu) [inline util/types/json/json.go]: I'll try this and compare them.

File: util/types/json/path.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (hanfei1991) [inline util/types/json/path.go]: also remove this file ?

**Comment by shenli** [comment]
Will we change the JSON binary representation in the future?

File: util/types/json/serdes.go [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+	"bytes"
+	"encoding/binary"
+	"reflect"
+	"sort"
+	"unsafe"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/hack"
+)
+
```
Comment (shenli) [inline util/types/json/serdes.go]: Could you add more comments and unit test for this file?

Comment (hicqu) [inline util/types/json/serdes.go]: After compare bson and mysql binary representation, I think the last one is good enough. But if we need, it's easy for adding a prefix flag to indicate which serialize format we use.

File: util/codec/codec.go [hunk 11]
```diff
@@ -202,6 +208,12 @@ func DecodeOne(b []byte) (remain []byte, d types.Datum, err error) {
 			v := types.Duration{Duration: time.Duration(r), Fsp: types.MaxFsp}
 			d.SetValue(v)
 		}
+	case jsonFlag:
+		var v []byte
+		var j json.JSON = json.CreateJSON(nil)
```
Comment (AndreMouche) [inline util/codec/codec.go]: I prefer `j := json.CreateJSON(nil)`

File: util/types/json/json.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	// ParseFromString parses a json from string.
+	ParseFromString(s string) error
+	// DumpToString dumps itself to string.
+	DumpToString() string
```
Comment (shenli) [inline util/types/json/json.go]: ToString is better?

File: util/types/json/json.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+// JSON is for MySQL JSON type.
+type JSON interface {
+	getTypeCode() byte
+	writeBinaryRepresentation(buffer *bytes.Buffer)
+	readBinaryRepresentation(data []byte) error
+}
+
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (JSON, error) {
+	return parseFromString(s)
```
Comment (hanfei1991) [inline util/types/json/json.go]: Why not implement `ParseFromString` directly ?

File: util/types/json/json.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// DumpToString dumps itself to string.
+func DumpToString(j JSON) string {
+	return dumpToString(j)
```
Comment (hanfei1991) [inline util/types/json/json.go]: Ditto

File: util/types/json/json.go [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"bytes"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/terror"
+)
+
+// JSON is for MySQL JSON type.
+type JSON interface {
+	getTypeCode() byte
+	writeBinaryRepresentation(buffer *bytes.Buffer)
```
Comment (shenli) [inline util/types/json/json.go]: encode/decode is more clear.

File: util/types/json/json.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	readBinaryRepresentation(data []byte) error
+}
+
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (JSON, error) {
+	return parseFromString(s)
+}
+
+// DumpToString dumps itself to string.
+func DumpToString(j JSON) string {
```
Comment (shenli) [inline util/types/json/json.go]: Can we implement Stringer interface for JSON?

File: util/types/json/serdes.go [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+	"github.com/pingcap/tidb/util/hack"
+)
+
+/*
+   The binary jSON format from MySQL 5.7 is as follows:
+
+   JSON doc ::= type value
+   type ::=
+       0x01 |       // large JSON object
```
Comment (shenli) [inline util/types/json/serdes.go]: Could we make those type as const?

File: util/types/json/serdes.go [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+
+   data-length ::= uint8*    // If the high bit of a byte is 1, the length
+                             // field is continued in the next byte,
+                             // otherwise it is the last byte of the length
+                             // field. So we need 1 byte to represent
+                             // lengths up to 127, 2 bytes to represent
+                             // lengths up to 16383, and so on...
+*/
+
+type jsonLiteral byte
```
Comment (hanfei1991) [inline util/types/json/serdes.go]: add the following code on the top of the file.
```
var (
_ JSON = new(jsonLiteral)
_ JSON = new(jsonString)
....
)
```

File: util/types/json/serdes.go [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+
+func (f jsonDouble) getTypeCode() byte {
+	return 0x0b
+}
+
+func (f jsonDouble) writeBinaryRepresentation(buffer *bytes.Buffer) {
+	binary.Write(buffer, binary.LittleEndian, f)
+}
+
+func (f *jsonDouble) readBinaryRepresentation(data []byte) error {
```
Comment (hanfei1991) [inline util/types/json/serdes.go]: Why do this use pointer ?

File: util/types/json/impl.go [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	return normalize(in), nil
+}
+
+func dumpToString(j JSON) string {
+	bytes, _ := json.Marshal(j)
+	return strings.Trim(string(bytes), "\n")
+}
+
+func normalize(in interface{}) JSON {
```
Comment (shenli) [inline util/types/json/impl.go]: Why we need to do this?

File: util/types/json/impl.go [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+		var literal = new(jsonLiteral)
+		if t {
+			*literal = jsonLiteral(0x01)
+		} else {
+			*literal = jsonLiteral(0x02)
+		}
+		return literal
+	case nil:
+		var literal = new(jsonLiteral)
+		*literal = 0x00
```
Comment (shenli) [inline util/types/json/impl.go]: Use const for those 0xxx.

File: util/types/json/impl.go (hunk 20 above)
Comment (hicqu) [inline util/types/json/impl.go]: we explicit implement `JSON` for `map[string]JSON` but not `map[string]interface{}`, so we need convert interface{} to JSON.

File: util/types/json/json.go (hunk 13 above)
Comment (hicqu) [inline util/types/json/json.go]: I think we shouldn't let interface level know these details.

File: util/types/json/impl.go [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+		return nil, ErrInvalidJSONText.GenByArgs("The document is empty")
+	}
+	var in interface{}
+	if err := json.Unmarshal([]byte(s), &in); err != nil {
+		return nil, ErrInvalidJSONText.GenByArgs(err)
+	}
+	return normalize(in), nil
+}
+
+func dumpToString(j JSON) string {
```
Comment (AndreMouche) [inline util/types/json/impl.go]: Any test for `ParseFromString ` and `DumpToString `?

File: util/types/json/json.go (hunk 13 above)
Comment (AndreMouche) [inline util/types/json/json.go]: I prefer to implement `ParseFromString` directly

**Comment by hicqu** [comment]
address comments.
@shenli @AndreMouche @hanfei1991 , PTAL again, thanks!

File: server/util.go [hunk 23]
```diff
@@ -348,6 +349,8 @@ func dumpTextValue(mysqlType uint8, value types.Datum) ([]byte, error) {
 		return hack.Slice(value.GetMysqlEnum().String()), nil
 	case types.KindMysqlSet:
 		return hack.Slice(value.GetMysqlSet().String()), nil
+	case types.KindMysqlJSON:
+		return hack.Slice(json.DumpToString(value.GetMysqlJSON())), nil
```
Comment (AndreMouche) [inline server/util.go]: Could we imitate above style and implement the function `String()` for `json.JSON`?

File: util/types/json/impl.go [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+		return []byte("true"), nil
+	default:
+		return []byte("false"), nil
+	}
+}
+
+// String implements JSON interface.
+func (j jsonLiteral) String() string {
+	bytes, _ := json.Marshal(j)
+	return strings.Trim(string(bytes), "\n")
```
Comment (AndreMouche) [inline util/types/json/impl.go]: Why not use `TrimSpace` directly?

File: util/types/json/impl.go [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+		return nil, ErrInvalidJSONText.GenByArgs(err)
+	}
+	return normalize(in), nil
+}
+
+func normalize(in interface{}) JSON {
+	switch t := in.(type) {
+	case bool:
+		if t {
+			return jsonLiteral(0x01)
```
Comment (AndreMouche) [inline util/types/json/impl.go]: please create a new constant to save `0x01 `

File: util/types/json/json.go [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package json
+
+import (
+	"bytes"
+	"github.com/pingcap/tidb/mysql"
```
Comment (AndreMouche) [inline util/types/json/json.go]: Please add a blank line before. You can find the `import` style from other source files.

**Comment by shenli** [comment]
@hicqu CI failed.
@AndreMouche PTAL

File: util/types/json/serdes.go [hunk 27]
```diff
@@ -0,0 +1,412 @@
...
+	// object ::= element-count size key-entry* value-entry* key* value*
+	var countAndSize = make([]uint32, 2)
+	var countAndSizeLen = len(countAndSize) * 4
+	var keySlice = keysInMap(m)
+
+	var keyEntrysLen = (4 + 2) * len(m)
+	var valueEntrysLen = (1 + 4) * len(m)
+	var keyEntrys = new(bytes.Buffer)
+	var valueEntrys = new(bytes.Buffer)
+	var keys = new(bytes.Buffer)
```
Comment (hicqu) [inline util/types/json/serdes.go]: I pefer short names, because keyBuffer need keyEntrys and keyEntrysLen respectively, but they are too long.

**Comment by hanfei1991** [comment]
LGTM

File: util/codec/codec.go [hunk 28]
```diff
@@ -202,6 +208,14 @@ func DecodeOne(b []byte) (remain []byte, d types.Datum, err error) {
 			v := types.Duration{Duration: time.Duration(r), Fsp: types.MaxFsp}
 			d.SetValue(v)
 		}
+	case jsonFlag:
+		var v []byte
+		if b, v, err = DecodeCompactBytes(b); err == nil {
+			var j json.JSON
+			if j, err = json.Deserialize(v); err == nil {
+				d.SetValue(j)
```
Comment (shenli) [inline util/codec/codec.go]: d.SetMysqlJSON(j) is better.

File: util/types/datum.go [hunk 29]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/util/charset"
 	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/types/json"
```
Comment (shenli) [inline util/types/datum.go]: Add some unit tests for the newly added functions.

File: util/types/json/impl.go [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/pingcap/tidb/util/hack"
+)
+
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (JSON, error) {
```
Comment (shenli) [inline util/types/json/impl.go]: Any unit tests for this function?

**Comment by shenli** [comment]
LGTM

**Pull request merged**